<table style="margin: 0 auto;">
  <tr>
    <td>
      <img src="https://www.genome.gov/sites/default/files/media/images/2019-03/tcga%20copy2.png"
           width="200"
           alt="TCGA logo">
    </td>
    <td style="padding-left: 20px;">
      <h1>Clinical Data Characterization</h1>
      <p>
        <strong>Ph.D. Marta Barea-Sepúlveda</strong><br>
        AI/ML Engineer &amp; Co-founder at Valendra
      </p>
    </td>
  </tr>
</table>

<style>
ol {
  counter-reset: item;
  padding-left: 20px;
}

ol li {
  display: block;
}

ol li::before {
  content: counters(item, ".") ". ";
  counter-increment: item;
  font-weight: bold;
}
</style>

<h2>Table of Contents</h2>

<hr>

<ol>
  <li><a href="#1-introduction">Introduction</a></li>

  <li><a href="#2-objectives">Objectives</a></li>

  <li>
    <a href="#3-environment-setup">Environment setup</a>
    <ol>
      <li><a href="#31-python-version">Python version</a></li>
      <li><a href="#32-required-libraries">Required libraries</a></li>
      <li><a href="#33-root-directory">Root directory</a></li>
    </ol>
  </li>

  <li><a href="#4-clinical-data-acquisition">Clinical data acquisition</a></li>

  <li><a href="#5-data-inspection-and-preparation">Data inspection and preparation</a>
      <ol>
          <li><a href="#51-raw-data-exploration">Raw data exploration</a>
          </li>
          <li><a href="#52-data-preparation">Data preparation</a></li>
          <li><a href="#53-prepared-data-overview">Prepared data overview</a></li>
        </ol>
  
  </li>
</ol>

<hr>

## 1. Introduction

<div align="justify">

This Jupyter Notebook is part of the **Disease Biology** phase of the project *"Integrative Multi-Omics Analysis for Drug Discovery in Clear Cell Renal Cell Carcinoma (ccRCC)"*. Specifically, it addresses the **Clinical Data Characterization** stage, focusing on the exploration and assessment of the clinical data available in the **[TCGA-KIRC (Kidney Renal Clear Cell Carcinoma)](https://portal.gdc.cancer.gov/projects/TCGA-KIRC)** cohort. 

TCGA-KIRC is part of **The Cancer Genome Atlas (TCGA)** program, a landmark cancer genomics initiative jointly supported by the **[National Cancer Institute (NCI)](https://www.cancer.gov/)** and the **[National Human Genome Research Institute (NHGRI)](https://www.genome.gov/)**.

</div>

## 2. Objetives

<div align="justify">

The objectives of this notebook are to review and prepare the downloaded clinical datasets, assess data completeness and quality, and characterize the clinical features of the patient cohort. 

To achieve this, a comprehensive statistical analysis will be conducted, including descriptive summaries of demographic and pathological variables, identification of missing data patterns, and evaluation of key clinical attributes. In addition, survival analyses will be performed to investigate the prognostic relevance of selected clinical variables and provide insights into patient outcomes.

</div>

## 3. Environment setup

### 3.1. Python version

In [21]:
# Show Python version
import sys
print(sys.version)

3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:54:21) [Clang 16.0.6 ]


### 3.2. Required libraries

In [ ]:
# Import necessary libraries
from pathlib import Path
import pandas as pd

### 3.3. Root directory 

In [23]:
# Root directory of the project
root_dir = Path.cwd().parents[1]

# Add project root to Python path
sys.path.append(str(root_dir))

## 4. Clinical data acquisition

<div align="justify">

The clinical dataset corresponding to the **[TCGA-KIRC](https://portal.gdc.cancer.gov/projects/TCGA-KIRC)** cohort was obtained from the **[Genomic Data Commons (GDC)](https://portal.gdc.cancer.gov)**, a platform maintained by the **[National Cancer Institute (NCI)](https://www.cancer.gov/)** that stores, harmonizes, and distributes genomic and clinical data from multiple research projects, including **The Cancer Genome Atlas (TCGA)**.

Clinical data were downloaded through the **[GDC API](https://gdc.cancer.gov/developers/gdc-application-programming-interface-api)**, which provides programmatic access to TCGA and other datasets hosted by the **[GDC](https://portal.gdc.cancer.gov)**. API requests were implemented in Python using HTTP requests to the appropriate API endpoints, applying filters to retrieve only the relevant clinical information. The implementation of this process is available in [`scripts/clinical_data_downloader.py`](../../scripts/clinical_data_downloader.py).

The downloaded clinical data are stored locally in the `data/clinical/` directory in TSV format. This directory is excluded from the repository via `.gitignore` to avoid redistributing data obtained from external sources. The files can be regenerated at any time by executing [`clinical_data_downloader.py`](../../scripts/clinical_data_downloader.py).

</div>

<div align="justify">

> For additional information about the **[GDC REST API](https://gdc.cancer.gov/developers/gdc-application-programming-interface-api)**, including query construction, available endpoints, and request parameters, please refer to the official **[GDC documentation](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/)**.

</div>

## 5. Data inspection and preparation

### 5.1. Raw data exploration

#### 5.1.1. Loading the clinical data

In [24]:
# Load TCGA KIRC clinical data
clinical_data = pd.read_csv(
    root_dir / "data" / "clinical" / "tcga_kirc_clinical.tsv",
    sep="\t"
)

#### 5.1.2. Data preview

In [30]:
# Display the first few rows of the clinical data
display(clinical_data.head())

,demographic.days_to_birth,demographic.days_to_death,demographic.gender,demographic.vital_status,diagnoses.0.ajcc_pathologic_m,diagnoses.0.ajcc_pathologic_n,diagnoses.0.ajcc_pathologic_stage,diagnoses.0.ajcc_pathologic_t,diagnoses.0.days_to_last_follow_up,diagnoses.0.laterality,...,diagnoses.1.tumor_grade,diagnoses.2.ajcc_pathologic_m,diagnoses.2.ajcc_pathologic_n,diagnoses.2.ajcc_pathologic_stage,diagnoses.2.ajcc_pathologic_t,diagnoses.2.days_to_last_follow_up,diagnoses.2.laterality,diagnoses.2.tumor_grade,id,submitter_id
0,-22320.0,NaN,female,Alive,M0,N0,Stage I,T1b,2150.0,Left,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,fb9bafa5-7133-4955-8156-4eb6763dc8e1,TCGA-B0-5695
1,-15360.0,NaN,male,Alive,NaN,NaN,NaN,NaN,NaN,Bilateral,...,G3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,33cf0893-0411-4758-aa90-602bfedf0850,TCGA-BP-4807
2,-24850.0,NaN,male,Alive,M0,N0,Stage I,T1b,1371.0,Right,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9245a557-01ea-4446-b9e9-313f6ab18834,TCGA-BP-4995
3,-15264.0,NaN,male,Alive,M0,NX,Stage I,T1a,1130.0,Right,...,NaN,MX,NX,Stage I,T1a,NaN,Left,NaN,c6454371-1f84-4e8e-8c3a-2eac8380819e,TCGA-DV-5573
4,-25508.0,1714.0,female,Dead,M1,N1,Stage IV,T4,1714.0,Left,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,33cd8b05-2b9b-40dd-97a2-40d02e27ee90,TCGA-CJ-4900


#### 5.1.3. Dataset schema

In [26]:

# Display information about the clinical data
clinical_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 537 entries, 0 to 536
Data columns (total 27 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   demographic.days_to_birth           536 non-null    float64
 1   demographic.days_to_death           177 non-null    float64
 2   demographic.gender                  537 non-null    str    
 3   demographic.vital_status            537 non-null    str    
 4   diagnoses.0.ajcc_pathologic_m       524 non-null    str    
 5   diagnoses.0.ajcc_pathologic_n       527 non-null    str    
 6   diagnoses.0.ajcc_pathologic_stage   522 non-null    str    
 7   diagnoses.0.ajcc_pathologic_t       526 non-null    str    
 8   diagnoses.0.days_to_last_follow_up  513 non-null    float64
 9   diagnoses.0.laterality              534 non-null    str    
 10  diagnoses.0.tumor_grade             510 non-null    str    
 11  diagnoses.1.ajcc_pathologic_m       27 non-null     str 

<div align="justify">

As shown in the dataset summary, the clinical dataset comprises **537 observations**, each corresponding to an individual patient, and **27 variables** containing demographic, clinical, and patient identification information.

Regarding data types, the dataset consists of **22 categorical variables** stored as `string` and **5 numerical variables** stored as `float64`. The numerical variables primarily represent time-related measurements, such as days to birth, days to death, and days to last follow-up.

The dataset also contains missing values. While some variables, such as `demographic.days_to_death`, are expected to have missing entries because the information is only available for deceased patients, a substantial proportion of missing values is observed in the variables prefixed with `diagnoses.1.*` and `diagnoses.2.*`. This is a consequence of the data structure provided by the **GDC**, which stores all available diagnosis records for each patient. The `diagnoses.0.*` variables correspond to the primary diagnosis and are available for nearly all patients, whereas the `diagnoses.1.*` and `diagnoses.2.*` variables represent additional diagnosis records. Since only a small subset of patients has multiple diagnosis entries, these variables are mostly empty and therefore exhibit a high percentage of missing values.


</div>

### 5.2. Data preparation

<div align="justify">

Based on the previous inspection, the clinical dataset was preprocessed to obtain a clean and consistent set of variables suitable for downstream analyses. The complete preprocessing pipeline is implemented in [`src/preprocessing/prepare_clinical_data.py`](../../src/preprocessing/prepare_clinical_data.py).

The main preprocessing steps are summarized below:

- **Reorganizing the dataset.** The patient identifiers (`id` and `submitter_id`) were moved to the first columns of the dataset to facilitate subsequent integration with other data modalities.

- **Retaining only the primary diagnosis.** Since the **[GDC](https://portal.gdc.cancer.gov)** stores multiple diagnosis records using the prefixes `diagnoses.0.*`, `diagnoses.1.*`, and `diagnoses.2.*`, only the primary diagnosis (`diagnoses.0.*`) was retained. The remaining diagnosis records correspond to a small subset of patients and contain a high proportion of missing values. After filtering, the common `diagnoses.0.` prefix was removed from the column names to improve readability.

- **Transforming clinical variables.** Patient age at diagnosis was computed from `days_to_birth`, which is stored by TCGA as the number of days before birth and therefore requires negation. In addition, the variables `days_to_last_follow_up` and `days_to_death` were converted from days to years.

- **Creating survival variables.** Two new variables were generated to support survival analysis:
  - `event`, a binary variable indicating whether the patient was deceased (`1`) or censored (`0`).
  - `survival_time`, defined as the time to death for deceased patients and the time to the last follow-up for censored patients.

- **Handling missing values.** Different strategies were applied depending on the variable:
  - Patients with missing values in `survival_time` or `ajcc_pathologic_stage` were removed, as these variables are essential for the subsequent analyses.
  - Missing values in the TNM staging variables (`ajcc_pathologic_t`, `ajcc_pathologic_n`, and `ajcc_pathologic_m`) were replaced with the corresponding unknown categories (`TX`, `NX`, and `MX`).
  - Missing values in `tumor_grade` were replaced with the unknown category `GX`.
  - Missing values in `laterality` were imputed using the mode.
  - Missing values in `age_at_diagnosis` were imputed using the cohort median.

As a result of these preprocessing steps, the clinical dataset contains standardized variables, harmonized time measurements, and complete information for all variables required in the downstream survival analyses.

</div>

### 5.3. Prepared data overview

#### 5.3.1. Loading the prepared clinical data

In [27]:
# Import prepared clinical data
from src.preprocessing.prepare_clinical_data import prepare_clinical_data

prepared_clinical_data = prepare_clinical_data()

#### 5.3.2. Data preview

In [ ]:
# Display the first few rows of the clinical data
display(prepared_clinical_data.head())

,id,submitter_id,gender,vital_status,ajcc_pathologic_m,ajcc_pathologic_n,ajcc_pathologic_stage,ajcc_pathologic_t,laterality,tumor_grade,age_at_diagnosis,years_to_last_follow_up,years_to_death,event,survival_time
0,fb9bafa5-7133-4955-8156-4eb6763dc8e1,TCGA-B0-5695,female,Alive,M0,N0,Stage I,T1b,Left,G2,61.108830,5.886379,NaN,0,5.886379
2,9245a557-01ea-4446-b9e9-313f6ab18834,TCGA-BP-4995,male,Alive,M0,N0,Stage I,T1b,Right,G3,68.035592,3.753593,NaN,0,3.753593
3,c6454371-1f84-4e8e-8c3a-2eac8380819e,TCGA-DV-5573,male,Alive,M0,NX,Stage I,T1a,Right,G2,41.790554,3.093771,NaN,0,3.093771
4,33cd8b05-2b9b-40dd-97a2-40d02e27ee90,TCGA-CJ-4900,female,Dead,M1,N1,Stage IV,T4,Left,G4,69.837098,4.692676,4.692676,1,4.692676
5,38469030-8539-4c76-b53b-57073569501c,TCGA-BP-4795,female,Alive,M0,N0,Stage I,T1a,Right,G2,74.365503,1.697467,NaN,0,1.697467


#### 5.3.3. Dataset schema

In [29]:
# Display information about the clinical data
prepared_clinical_data.info()

<class 'pandas.DataFrame'>
Index: 512 entries, 0 to 536
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       512 non-null    str    
 1   submitter_id             512 non-null    str    
 2   gender                   512 non-null    str    
 3   vital_status             512 non-null    str    
 4   ajcc_pathologic_m        512 non-null    str    
 5   ajcc_pathologic_n        512 non-null    str    
 6   ajcc_pathologic_stage    512 non-null    str    
 7   ajcc_pathologic_t        512 non-null    str    
 8   laterality               512 non-null    str    
 9   tumor_grade              512 non-null    str    
 10  age_at_diagnosis         512 non-null    float64
 11  years_to_last_follow_up  510 non-null    float64
 12  years_to_death           170 non-null    float64
 13  event                    512 non-null    int64  
 14  survival_time            512 non-null    f

<div align="justify">

After applying the preprocessing pipeline, the clinical dataset contains **512 patients** and **15 variables**, representing a reduction from the original dataset due to the removal of patients with missing survival information or pathological stage, as these variables are essential for the subsequent analyses.

The resulting dataset includes patient identifiers, demographic variables, pathological staging information, tumor grade, laterality, and the derived survival variables (`event` and `survival_time`). Most of the original missing values have been addressed through the preprocessing strategy described in the previous section.

Regarding data types, the final dataset consists of **10 categorical variables** (`string`), **4 continuous numerical variables** (`float64`), and **1 binary variable** (`int64`). The only remaining missing values correspond to `years_to_last_follow_up` and `years_to_death`, which are expected because these variables are mutually exclusive. Patients who are alive have no recorded time to death, whereas deceased patients have no follow-up time after death. Consequently, the derived variable `survival_time` contains complete information for all 512 patients and can be directly used for downstream survival analyses.

</div>